### ISSUE TITLE
Get Grain from generic grain id through IGrainFactory

### ISSUE THREAD

**Issue by nimashoghi** [dotnet/orleans#5541]
I am currently developing an abstract base class for a set of worker grains that operate on a parent grain. I want to create a way to get the parent grain.

Here's a simple exmaple of what I'm doing:
```fs
[<AbstractClass>]
type GrainWorker<'grain when 'grain :> IGrain> () =
    inherit Grain ()

    member this.ParentGrain () = this.GrainFactory.GetGrain<'grain> (this.GetPrimaryKey ())

type IUserGrain =
    inherit IGrainWithGuidKey

type IUserWorkerGrain =
    inherit IGrainWithGuidKey

type UserGrainWorker () =
    inherit GrainWorker<IUserGrain> ()
```
Note that in the scenario above, `this.GetPrimaryKey ()` returns a Guid key, so GrainWorker would only work for Guid grains. I want to have a way to do this operation in a generic manner, without having to duplicate tons of code.

Based off of my minor research, it seems like the IInternalGrainFactory methods would be useful in this case, but this interface is marked as internal.

As @ReubenBond noted in the Gitter chat, issue #1511 is related to this.

**Comment by ReubenBond** [comment]
@nimashoghi are you looking for a way to get a grain from the grain factory using `IGrainIdentity`? We could accept a PR for that. I just closed #1511, since #5674 implements what it is asking for.

**Comment by ReubenBond** [comment]
Fixed in 4.0: `IGrainFactory` now has overloads accepting `GrainId` (which is now a public type) and returning untyped grain references (which can be converted into typed references using 's.AsReference<T>()`)
